### PR TITLE
compiletest: error when finding a trailing directive

### PR DESCRIPTION
**Pull request by Urgau**
This PR introduce a supplementary check that when checking for a compiletest directive, will also check that the next "word" after that directive is not also by itself a directive.

This is done to avoid situations like this `//@ only-linux only-x86_64` where one might think that both directives are being taken into account while in fact the second in a comment, and so was ignored, until now.

Related to #123730
cc @scottmcm
r? @jieyouxu

### CODE REVIEW

File: src/tools/compiletest/src/header.rs [hunk 1]
```diff
@@ -935,16 +935,26 @@ struct HeaderLine<'ln> {
 pub(crate) struct CheckDirectiveResult<'ln> {
     is_known_directive: bool,
     directive_name: &'ln str,
+    trailing_directive: Option<&'ln str>,
 }
 
 // Returns `(is_known_directive, directive_name)`.
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:941]: This comment is outdated, probably just delete it.

File: src/tools/compiletest/src/header.rs [hunk 2]
```diff
@@ -1028,6 +1039,19 @@ fn iter_header(
 
                     return;
                 }
+
+                if let Some(trailing_directive) = &trailing_directive {
+                    *poisoned = true;
+
+                    eprintln!(
+                        "error: detected trailing compiletest test directive `{}` in {}:{}",
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:1047]: At the cost of verbosity, should we also recommend what the user should do, i.e. some wording like "one directive per line"? Or maybe reword this as something like "each line can only contain one compiletest directive"?

File: src/tools/compiletest/src/header.rs [hunk 3]
```diff
@@ -1051,7 +1075,8 @@ fn iter_header(
 
             let rest = rest.trim_start();
 
-            let CheckDirectiveResult { is_known_directive, directive_name } = check_directive(rest);
+            let CheckDirectiveResult { is_known_directive, directive_name, trailing_directive: _ } =
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:1078]: Probably just

```suggestion
            let CheckDirectiveResult { is_known_directive, directive_name, .. } =
```

if we're not going to use the trailing directive.

File: src/tools/compiletest/src/header/tests.rs [hunk 4]
```diff
@@ -667,3 +667,24 @@ fn test_non_rs_unknown_directive_not_checked() {
...
+fn test_trailing_directive_with_comment() {
+    let mut poisoned = false;
+    run_path(&mut poisoned, Path::new("a.rs"), b"//@ only-x86   only-arm with comment");
+    assert!(poisoned);
+}
+
+#[test]
+fn test_not_trailing_directive() {
+    let mut poisoned = false;
+    run_path(&mut poisoned, Path::new("a.rs"), b"//@ revisions: incremental");
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header/tests.rs:688]: would this reject `//@ revisions incremental`? I want to say that the `:` was not enforced in compiletest directive grammar when parsing directives...

**Review by jieyouxu — COMMENTED** [review]
Changes look mostly good to me, just some nitpicks and a question regarding if this would reject e.g. `//@ revisions incremental` (without the `:`).

File: src/tools/compiletest/src/header/tests.rs (hunk 4 above)
Comment (Urgau) [reply src/tools/compiletest/src/header/tests.rs:688]: No it would not be rejected, because it's currently used in `tests/ui/invalid-compile-flags/fuel.rs`:
https://github.com/rust-lang/rust/blob/b3bd7058c139e71bae0862ef8f8ac936208873e9/tests/ui/invalid-compile-flags/fuel.rs#L1

I also don't see a reason to reject it, since the `:` is a delimiter.

Comment (jieyouxu) [reply src/tools/compiletest/src/header/tests.rs:688]: Sorry let me clarify what I mean. I believe compiletest accepts both `//@ revisions incremental` and `//@ revisions: incremental`, would the logic in this PR (incorrectly) reject `//@ revisions incremental`?

File: src/tools/compiletest/src/header.rs (hunk 1 above)
Comment (Urgau) [reply src/tools/compiletest/src/header.rs]: Sure, removed.

File: src/tools/compiletest/src/header.rs (hunk 2 above)
Comment (Urgau) [reply src/tools/compiletest/src/header.rs]: I've added a help line.

```
error: detected trailing compiletest test directive `only-arm` in a.rs:1
 help: put the trailing directive in it's own line: `//@ only-arm`
```

File: src/tools/compiletest/src/header/tests.rs (hunk 4 above)
Comment (jieyouxu) [reply src/tools/compiletest/src/header/tests.rs:688]: Yeah, unfortunately,

```
//@ revisions incremental
fn main() {}
```

would get rejected

```
error: detected trailing compiletest test directive `incremental`
```

whereas

```
//@ revisions: incremental
fn main() {}
```

the thing here is that compiletest accepts *both* forms (unfortunate directive grammar strikes yet again)

File: src/tools/compiletest/src/header.rs (hunk 2 above)
Comment (jieyouxu) [reply src/tools/compiletest/src/header.rs]: Hmm, I wonder if we should say "each line can only contain one compiletest directive" instead of saying both trailing detected + the help, because if someone writes `//@ only-x64 only-arm only-wasm`, then they follow this help, but the help contains `//@ only-arm only-wasm` (which is not correct if we reject trailing directives again).

File: src/tools/compiletest/src/header/tests.rs (hunk 4 above)
Comment (Urgau) [reply src/tools/compiletest/src/header/tests.rs:688]: No, no, it's me. I read the question the wrong way round.

> I believe compiletest accepts both `//@ revisions incremental` and `//@ revisions: incremental`

I just checked, and it doesn't seems to be the case. There is no revisions (also no errors).

> would the logic in this PR (incorrectly) reject `//@ revisions incremental`?

Yes, it would be rejected by this PR; but I would also argue that it's not incorrect since the syntax is a nop and should probably be an error.

Comment (jieyouxu) [reply src/tools/compiletest/src/header/tests.rs:688]: > I just checked, and it doesn't seems to be the case. There is no revisions (also no errors).

**What**

Comment (jieyouxu) [reply src/tools/compiletest/src/header/tests.rs:688]: Ah darn, I think it's because of a combination of things:

1. `revisions` is a known compiletest directive
2. when we parse name-value directives, the directive name is *required* to have a trailing `colon` in order for the parse to succeed: https://github.com/rust-lang/rust/blob/b14d8b2ef20c64c1002e2c6c724025c3d0846b91/src/tools/compiletest/src/header.rs#L1137-L1146

**However**, the way compiletest directive is setup right now means that if known directive check allows a directive but the corresponding name-value directive parsing logic fails, we **silently** accept the directive but **it has no effect**.

This is awful LOL

Comment (jieyouxu) [reply src/tools/compiletest/src/header/tests.rs:688]: Given that the colon is actually mandatory for directives like `revisions`, I think it's okay to reject things like `//@ only-arch known-directive` because the trailing known-directive-looking-thing is very unlikely to be a directive commment.

File: src/tools/compiletest/src/header.rs (hunk 2 above)
Comment (Urgau) [reply src/tools/compiletest/src/header.rs]: Except that I specifically made sure that things like that `//@ only-x64 only-arm only-wasm` would only check `only-arm` and thus the help would only suggest: `//@ only-arm`

File: src/tools/compiletest/src/header/tests.rs (hunk 4 above)
Comment (jieyouxu) [reply src/tools/compiletest/src/header/tests.rs:688]: I filed an issue https://github.com/rust-lang/rust/issues/123760 to make sure we know that this is a bug, but it's out of the scope for this PR. I'm happy to accept this PR as is.

**Comment by jieyouxu** [comment]
Changes look good to me and I think it's okay to reject `//@ <known_directive_name> <another_known_directive_name>` given that `<another_known_directive_name>` as a single word without `-` is rare (e.g. `incremental` or `edition`), and for directive names that does have a `-`, it's very unlikely that a directive comment would start with a known directive name.

Feel free to r=me after CI is green.

@bors delegate+
